wallet-new/test/unit/TxMetaStorageSpecs.hs

```diff
@@ -0,0 +1,167 @@
+               catchJust (\e -> if isDoesNotExistError e then Just () else Nothing)
+                         (removeFile "txmeta_test.db")
+                         (\_ -> return ())
+               db <- openMetaDB "txmeta_test.db"
+               Storage.unsafeMigrate db
+               return db
+
+       release :: MetaDBHandle -> m ()
+       release = liftIO . closeMetaDB
+
+
+-- | Generates two 'TxMeta' which are @almost@ identical, if not in the
+-- arrangement of their inputs & outputs.
+genSimilarTxMetas :: IO (TxMeta, TxMeta)
+genSimilarTxMetas = liftIO $ do
+    inputs  <- generate (resize 5 arbitrary)
+    outputs <- generate (resize 5 arbitrary)
+    blueprint <- generate arbitrary
+    let t1 = blueprint & over txMetaInputs  (const inputs)
+                       . over txMetaOutputs (const outputs)
+    let t2 = blueprint & over txMetaInputs  (const (NonEmpty.reverse inputs))
+                       . over txMetaOutputs (const (NonEmpty.reverse outputs))
+    return (t1, t2)
+
+
+-- | Handy wrapper to be able to compare things with the 'isomorphicTo'
+-- combinator, which ignores the different order of the inputs & outputs.
+data DeepEqual = DeepEqual TxMeta
+
+instance Eq DeepEqual where
+    (DeepEqual t1) == (DeepEqual t2) = t1 `exactlyEqualTo` t2
+
+instance Buildable DeepEqual where
+    build (DeepEqual t) = build t
+
+data Isomorphic = Isomorphic TxMeta
+
+instance Eq Isomorphic where
+    (Isomorphic t1) == (Isomorphic t2) = t1 `isomorphicTo` t2
+
+instance Buildable Isomorphic where
+    build (Isomorphic t) = build t
+
+instance Buildable [Isomorphic] where
+    build ts = bprint (listJsonIndent 4) ts
+
+sortByAmount :: SortDirection -> [Isomorphic] -> [Isomorphic]
+sortByAmount direction = sortBy sortFn
+    where
+        withDir Ascending  = identity
+        withDir Descending = flip
+
+        sortFn (Isomorphic a) (Isomorphic b) =
+            (withDir direction compare) (a ^. txMetaAmount) (b ^. txMetaAmount)
+
+sortByCreationAt :: SortDirection -> [Isomorphic] -> [Isomorphic]
+sortByCreationAt direction = sortBy sortFn
+    where
+        withDir Ascending  = identity
+        withDir Descending = flip
+
+        sortFn (Isomorphic a) (Isomorphic b) =
+            (withDir direction compare) (a ^. txMetaCreationAt) (b ^. txMetaCreationAt)
+
+-- | Specs which tests the persistent storage and API provided by 'TxMeta'.
+txMetaStorageSpecs :: Spec
+txMetaStorageSpecs = do
+    describe "TxMeta equality" $ do
+        it "should be reflexive" $ do
+            (blueprint :: TxMeta) <- liftIO $ generate arbitrary
+            blueprint `exactlyEqualTo` blueprint `shouldBe` True
+
+        it "should be strict when needed" $ do
+            (t1, t2) <- liftIO genSimilarTxMetas
+            t1 `exactlyEqualTo` t2 `shouldBe` False
+
+        it "isomorphicTo is more lenient" $ do
+            (t1, t2) <- liftIO genSimilarTxMetas
+            t1 `isomorphicTo` t2 `shouldBe` True
```

Comment:
likewise, 

```haskell
prop "isomoprhic is more leniant"
    $ forAll genSimilarTxMetas
    $ \(t1, t2) ->
        t1 `isomorphicTo` t2
```

The main reason we want this is because the *error* we get from failure will be much better. With your current code, we'll get something like:

```
isomorphic is more lenient FAILED
    False should be True
```

but using `prop` we'll get the actual values that failed the test.

Reply:
Additionally, the `prop` things will generate 100 or so eaxmples. `liftIO (generate arbitrary)` only tests a single thing!